### DISCUSSION TITLE
Can I convert stablehlo/mhlo into triton IR?

### DISCUSSION THREAD

**Discussion by syheliel** [openxla/community#102] [stablehlo]
I've a deep search on openxla repo. But I can only find the convert from hlo(the non-mlir one) into triton IR in openxla/xla. The code is [here](https://github.com/openxla/xla/blob/ad9b4948261f9cd048db15c2ff9fe8791f3c5a34/xla/service/gpu/ir_emitter_triton_test.cc#L181). 

Is there an existing codebase implements the convertion from stablehlo/mhlo int triton IR through MLIR?

![image](https://github.com/openxla/community/assets/45957390/e7b5d4a7-9a6e-4d3d-9df2-5ce0730a83d4)

**Comment by GleasonK** [comment]
I don't believe any e2e tool exists for this yet, but seems like it could be made, or even added to [`hlo-opt`](https://github.com/openxla/xla/tree/1b2dc850e1a9af62f600585eafef77b7102c49fa/xla/tools/hlo_opt) would likely be a welcome feature / contribution.

We have all the functions to bring StableHLO to HLO - PyTorch/XLA is probably the most concise example of using these conversions in [`ConvertStableHloToHlo`](https://github.com/pytorch/xla/blob/2e6e183e0724818f137c8135b34ef273dea33318/torch_xla/csrc/runtime/stablehlo_helper.cc#L156). From there it should be a matter of using the function you've pointed out above.

An alternative would be to use existing tools to convert StableHLO to HLO: [`mlir-hlo-opt --stablehlo-legalize-to-hlo`](https://github.com/openxla/xla/blob/1b2dc850e1a9af62f600585eafef77b7102c49fa/xla/mlir_hlo/tests/Dialect/mhlo/stablehlo-legalize-to-hlo.mlir#L1) and [`xla-translate --mlir-hlo-to-hlo-text`](https://github.com/openxla/xla/blob/1b2dc850e1a9af62f600585eafef77b7102c49fa/xla/translate/mhlo_to_hlo/tests/export.mlir#L1).

This all said - I'm not an expert in the triton emission pipeline. Will let someone else chime in on how ready this path is for general emission from StableHLO programs, etc.

**Comment by cheshire** [comment]
It's in the plans to add such functionality to the `hlo-opt` tool. Note that not all the fusions are convertible to TritonIR, so it would have to return a subset of the original HLO module (or error out if non-convertible fusions exist).
